ext/Client/VisualEnvironmentHandler.lua

```diff
@@ -0,0 +1,409 @@
+		local s_Object = self._VisualEnvironmentObjects[l_ID]
+
+		-- check if alive
+		if s_Object and s_Object.entity and s_Object.ve.priority == 1 then
+			self:DestroyVE(l_ID)
+			self._Lerping[l_ID] = {}
+		end
+	end
+end
+
+---@param p_DeltaTime number
+function VisualEnvironmentHandler:UpdateLerp(p_DeltaTime)
+	for l_ID, l_LerpingTable in pairs(self._Lerping) do
+		local s_TimeSinceStart = SharedUtils:GetTimeMS() - l_LerpingTable.startTime
+		local s_CompletionPercentage = s_TimeSinceStart / l_LerpingTable.transitionTime * 100
+
+		if l_LerpingTable.firstRun then
+			-- prevent destroying on first run
+			l_LerpingTable.startValue = math.max(l_LerpingTable.startValue, 0.01)
+			l_LerpingTable.firstRun = false
+		end
+
+		-- ! for now only use functions that need these parameters
+		-- t = elapsed time (ms)
+		-- b = begin
+		-- c = change == ending - beginning
+		-- d = duration (total time, ms)
+		local t = s_TimeSinceStart
+		local b = l_LerpingTable.startValue
+		local c = l_LerpingTable.endValue - l_LerpingTable.startValue
+		local d = l_LerpingTable.transitionTime
+
+		local s_TransitionFunction = l_LerpingTable.transitionFunction
+
+		---@type number
+		local s_LerpValue = s_TransitionFunction(t, b, c, d)
+
+		---@diagnostic disable-next-line: param-type-mismatch
+		if s_CompletionPercentage >= 100 then
+			if l_LerpingTable.pulsing then
+				l_LerpingTable.startTime = SharedUtils:GetTimeMS()
+
+				-- Swap if pulsing, so we set the opposite visibility as goal
+				if s_LerpValue < 0.1 then
+					l_LerpingTable.startValue = 0.0
+					l_LerpingTable.endValue = 1.0
+				else
+					l_LerpingTable.startValue = 1.0
+					l_LerpingTable.endValue = 0.0
+				end
+			else
+				self:SetVisibility(l_ID, l_LerpingTable.endValue)
+				self._Lerping[l_ID] = nil
+			end
+		elseif s_CompletionPercentage < 0 then
+			m_Logger:Warning('Lerping of preset ' .. tostring(l_ID) .. ' has its completed percentage of ' .. tostring(s_CompletionPercentage) .. ', should never happen')
+			self:SetVisibility(l_ID, l_LerpingTable.endValue)
+			self._Lerping[l_ID] = nil
+		else
+			self:SetVisibility(l_ID, s_LerpValue)
+		end
+	end
+end
+
+---@param p_ID string
+---@param p_Class string
+---@param p_Property string
+---@param p_Value any
+function VisualEnvironmentHandler:SetSingleValue(p_ID, p_Class, p_Property, p_Value)
+	if not p_Class or not p_Property or not p_Value then
+		m_Logger:Write("Passed invalid parameters")
+		return
+	end
+	---@type VisualEnvironmentObject
+	local s_Object = self._VisualEnvironmentObjects[p_ID]
+	s_Object.entity.state[p_Class][p_Property] = p_Value
+	VisualEnvironmentManager:SetDirty(true)
+end
+
+---@param p_ID string
+---@param p_Guid Guid
+---@param p_Path string
+function VisualEnvironmentHandler:ApplyTexture(p_ID, p_Guid, p_Path)
+	---@type VisualEnvironmentObject
+	local s_Object = self._VisualEnvironmentObjects[p_ID]
+
+	for _, l_Class in pairs(s_Object.ve.components) do
+
+		if l_Class.typeInfo.name == "SkyComponentData" then
```

Comment:
I think Nofate recommended to use `l_Class:Is('SkyComponentData')`, would be nice to change it while were at it

Reply:
I remembered it being the other way around :Is() being less performant but ok :D will do